src/main/java/com/rackspace/ceres/app/web/TsdbQueryController.java

```diff
@@ -0,0 +1,93 @@
+package com.rackspace.ceres.app.web;
+
+import com.rackspace.ceres.app.model.LookupResult;
+import com.rackspace.ceres.app.model.MetricNameAndMultiTags;
+import com.rackspace.ceres.app.model.SeriesData;
+import com.rackspace.ceres.app.services.MetadataService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Profile;
+import org.springframework.web.bind.annotation.*;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import java.util.*;
+
+/**
+ * Native Ceres search API endpoints.
+ */
+@RestController
+@RequestMapping("/api/search")
+@Profile("query")
+public class TsdbQueryController {
+    private final MetadataService metadataService;
+
+    @Autowired
+    public TsdbQueryController(MetadataService metadataService) {
+        this.metadataService = metadataService;
+    }
+
+    @GetMapping("/lookup")
+    public Mono<LookupResult> query(@RequestParam(name = "m") String m,
+                                    @RequestParam(name = "limit") Integer limit,
+                                    @RequestHeader(value = "X-Tenant", required = true) String tenantHeader) {
+
+        List<SeriesData> results = new ArrayList<SeriesData>();
+        final MetricNameAndMultiTags metricNameAndTags = getMetric(m);
+
+        return metadataService.getTagKeys(tenantHeader, metricNameAndTags.getMetricName())
```

Comment:
Can we create some flux utility functions to do most of this? Perhaps that should be a refactoring ticket all on its own that gets pulled into next sprint.

Reply:
Sounds like a good idea. Let's write up a JIRA.

Reply:
and please include moving utility methods from controller class to a service class in that JIRA ticket